src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp

```diff
@@ -33,63 +32,72 @@ WriteBufferFromAzureBlobStorage::~WriteBufferFromAzureBlobStorage()
     finalize();
 }
 
-void WriteBufferFromAzureBlobStorage::finalizeImpl()
+void WriteBufferFromAzureBlobStorage::execWithRetry(std::function<void()> func, size_t num_tries)
 {
-    if (attributes.has_value())
+    auto can_retry_exception = [&]<typename Exception>(const Exception & e, size_t i) -> bool
     {
-        auto blob_client = blob_container_client->GetBlobClient(blob_path);
-        Azure::Storage::Metadata metadata;
-        for (const auto & [key, value] : *attributes)
-            metadata[key] = value;
-        blob_client.SetMetadata(metadata);
-    }
+        if (i == num_tries - 1)
+            return false;
 
-    const size_t max_tries = 3;
-    for (size_t i = 0; i < max_tries; ++i)
+        LOG_DEBUG(log, "Write at attempt {} for blob `{}` failed: {}", i + 1, blob_path, e.Message);
+        return true;
+    };
+
+    for (size_t i = 0; i < num_tries; ++i)
     {
         try
         {
-            next();
+            func();
             break;
         }
+        catch (const Azure::Core::Http::TransportException & e)
+        {
+            if (!can_retry_exception(e, i))
+                throw;
+        }
         catch (const Azure::Core::RequestFailedException & e)
         {
-            if (i == max_tries - 1)
+            if (!can_retry_exception(e, i))
                 throw;
-            LOG_INFO(&Poco::Logger::get("WriteBufferFromAzureBlobStorage"),
-                     "Exception caught during finalizing azure storage write at attempt {}: {}", i + 1, e.Message);
         }
     }
 }
 
+void WriteBufferFromAzureBlobStorage::finalizeImpl()
+{
+    execWithRetry([this](){ next(); }, 3);
+}
+
 void WriteBufferFromAzureBlobStorage::nextImpl()
 {
     if (!offset())
         return;
 
-    auto * buffer_begin = working_buffer.begin();
-    auto len = offset();
+    char * buffer_begin = working_buffer.begin();
+    size_t total_size = offset();
+
     auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
 
-    size_t read = 0;
+    size_t current_size = 0;
     std::vector<std::string> block_ids;
-    while (read < len)
-    {
-        auto part_len = std::min(len - read, max_single_part_upload_size);
 
-        auto block_id = getRandomASCIIString(64);
-        block_ids.push_back(block_id);
+    while (current_size < total_size)
+    {
+        size_t part_len = std::min(total_size - current_size, max_single_part_upload_size);
+        const std::string & block_id = block_ids.emplace_back(getRandomASCIIString(64));
 
-        Azure::Core::IO::MemoryBodyStream tmp_buffer(reinterpret_cast<uint8_t *>(buffer_begin + read), part_len);
-        block_blob_client.StageBlock(block_id, tmp_buffer);
+        Azure::Core::IO::MemoryBodyStream tmp_buffer(reinterpret_cast<uint8_t *>(buffer_begin + current_size), part_len);
+        execWithRetry([&block_blob_client, &block_id, &tmp_buffer](){ block_blob_client.StageBlock(block_id, tmp_buffer); }, 3);
```

Comment:
```suggestion
        execWithRetry([&](){ block_blob_client.StageBlock(block_id, tmp_buffer); }, 3);
```
Why not this?
Only used variables will be captured.
For me, explicitly defining each variable is when you mix value and reference captures.